### PR TITLE
planner: fix panic when the join key is scalarFunction

### PR DESCRIPTION
**Pull request by Yisaer**
Signed-off-by: yisaer <disxiaofei@163.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/29401

Problem Summary:

If one side of the join key is string, and the another is json, the expression of the string would be cast to json and marked `ParseToJSONFlag` which will let `WrapWithCastAsJSON` wrap extra cast function.

In order to avoid this problem, we remove `ParseToJSONFlag` temporary during `updateEQCond`.

The final explain info result for the original issue is like following:

```mysql
mysql> explain select /*+ inl_hash_join(t1) */ * from t t1 join t t2 on t1.b=t2.e;
+-----------------------------+----------+-----------+---------------+---------------------------------------------------------------------------------------------------------------+
| id                          | estRows  | task      | access object | operator info                                                                                                 |
+-----------------------------+----------+-----------+---------------+---------------------------------------------------------------------------------------------------------------+
| HashJoin_8                  | 12500.00 | root      |               | inner join, equal:[eq(Column#15, test.t.e)]                                                                   |
| ├─TableReader_14(Build)     | 10000.00 | root      |               | data:TableFullScan_13                                                                                         |
| │ └─TableFullScan_13        | 10000.00 | cop[tikv] | table:t2      | keep order:false, stats:pseudo                                                                                |
| └─Projection_10(Probe)      | 10000.00 | root      |               | test.t.id, test.t.a, test.t.b, test.t.c, test.t.d, test.t.e, test.t.f, cast(test.t.b, json BINARY)->Column#15 |
|   └─TableReader_12          | 10000.00 | root      |               | data:TableFullScan_11                                                                                         |
|     └─TableFullScan_11      | 10000.00 | cop[tikv] | table:t1      | keep order:false, stats:pseudo                                                                                |
+-----------------------------+----------+-----------+---------------+---------------------------------------------------------------------------------------------------------------+
6 rows in set, 1 warning (0.00 sec)

```

### What is changed and how it works?

This request fixs this case.

### Check List

Tests 

- [x] Unit test

### Release note

```release-note
Fix the bug that sql got cancel if including json column joins char column.
```

### CODE REVIEW

**Review by winoros — COMMENTED** [review]
It's a little strange that we would add a projection if there's function in the equal condition.
Such as
```
mysql> explain select * from t1, t2 where t1.a+1=t2.a;
+--------------------------------+----------+-----------+---------------+---------------------------------------------+
| id                             | estRows  | task      | access object | operator info                               |
+--------------------------------+----------+-----------+---------------+---------------------------------------------+
| Projection_10                  | 12500.00 | root      |               | test.t1.a, test.t2.a                        |
| └─HashJoin_11                  | 12500.00 | root      |               | inner join, equal:[eq(test.t2.a, Column#5)] |
|   ├─Projection_15(Build)       | 10000.00 | root      |               | test.t1.a, plus(test.t1.a, 1)->Column#5     |
|   │ └─TableReader_17           | 10000.00 | root      |               | data:TableFullScan_16                       |
|   │   └─TableFullScan_16       | 10000.00 | cop[tikv] | table:t1      | keep order:false, stats:pseudo              |
|   └─TableReader_14(Probe)      | 10000.00 | root      |               | data:TableFullScan_13                       |
|     └─TableFullScan_13         | 10000.00 | cop[tikv] | table:t2      | keep order:false, stats:pseudo              |
+--------------------------------+----------+-----------+---------------+---------------------------------------------+
7 rows in set (0.01 sec)
```
In the above case, we add projection(t1.a+1=>Column#5) then let the #5 serve as the join key.

I think we should keep such hypothesis. Add projection when we add the cast.

**Comment by Yisaer** [comment]
The hypothesis is still kept. The generated explain info is like following:

```mysql
mysql> explain select /*+ inl_hash_join(t1) */ * from t t1 join t t2 on t1.b=t2.e;
+-----------------------------+----------+-----------+---------------+---------------------------------------------------------------------------------------------------------------+
| id                          | estRows  | task      | access object | operator info                                                                                                 |
+-----------------------------+----------+-----------+---------------+---------------------------------------------------------------------------------------------------------------+
| HashJoin_8                  | 12500.00 | root      |               | inner join, equal:[eq(cast(Column#15, json BINARY), test.t.e)]                                                |
| ├─TableReader_14(Build)     | 10000.00 | root      |               | data:TableFullScan_13                                                                                         |
| │ └─TableFullScan_13        | 10000.00 | cop[tikv] | table:t2      | keep order:false, stats:pseudo                                                                                |
| └─Projection_10(Probe)      | 10000.00 | root      |               | test.t.id, test.t.a, test.t.b, test.t.c, test.t.d, test.t.e, test.t.f, cast(test.t.b, json BINARY)->Column#15 |
|   └─TableReader_12          | 10000.00 | root      |               | data:TableFullScan_11                                                                                         |
|     └─TableFullScan_11      | 10000.00 | cop[tikv] | table:t1      | keep order:false, stats:pseudo                                                                                |
+-----------------------------+----------+-----------+---------------+---------------------------------------------------------------------------------------------------------------+
6 rows in set, 1 warning (0.01 sec)
```


The cast function was added by `newBaseBuiltinFuncWithTp` in `WrapWithCastAsJSON` during we build eq condition in [updateEqCond](https://github.com/pingcap/tidb/blob/9535f0592f6deb371b15d4746fd9dcc7b895c573/planner/core/rule_predicate_push_down.go#L276) and I'm not sure whether removing this wrap would be a good solution. @winoros

**Comment by winoros** [comment]
We found that in the explain, there's `cast(test.t.b, json BINARY)->Column#15` then `eq(cast(Column#15, json BINARY)`.
The test.t.b cast to json twice. This should be the problem.
Refer to the code
```
// WrapWithCastAsJSON wraps `expr` with `cast` if the return type of expr is not
// type json, otherwise, returns `expr` directly.
func WrapWithCastAsJSON(ctx sessionctx.Context, expr Expression) Expression {
	if expr.GetType().Tp == mysql.TypeJSON && !mysql.HasParseToJSONFlag(expr.GetType().Flag) {
		return expr
	}
	tp := &types.FieldType{
		Tp:      mysql.TypeJSON,
		Flen:    12582912, // FIXME: Here the Flen is not trusted.
		Decimal: 0,
		Charset: mysql.DefaultCharset,
		Collate: mysql.DefaultCollationName,
		Flag:    mysql.BinaryFlag,
	}
	return BuildCastFunction(ctx, expr, tp)
}
```
The `Column#15`'s `ParseToJSONFlag` is still true. So we do cast twice.
But I've roughly gone through all the related codes about how we build EQ function and how we add cast function. Surprisingly that I think that flag should be reset though it's not in fact.:(

**Comment by Yisaer** [comment]
after re-fixing, the result like following:

```mysql
mysql> explain select /*+ inl_hash_join(t1) */ * from t t1 join t t2 on t1.b=t2.e;
+------------------------------+----------+-----------+---------------+---------------------------------------------------------------------------------------------------------------+
| id                           | estRows  | task      | access object | operator info                                                                                                 |
+------------------------------+----------+-----------+---------------+---------------------------------------------------------------------------------------------------------------+
| HashJoin_9                   | 12500.00 | root      |               | inner join, equal:[eq(Column#15, Column#16)]                                                                  |
| ├─Projection_14(Build)       | 10000.00 | root      |               | test.t.id, test.t.a, test.t.b, test.t.c, test.t.d, test.t.e, test.t.f, cast(test.t.e, json BINARY)->Column#16 |
| │ └─TableReader_16           | 10000.00 | root      |               | data:TableFullScan_15                                                                                         |
| │   └─TableFullScan_15       | 10000.00 | cop[tikv] | table:t2      | keep order:false, stats:pseudo                                                                                |
| └─Projection_11(Probe)       | 10000.00 | root      |               | test.t.id, test.t.a, test.t.b, test.t.c, test.t.d, test.t.e, test.t.f, cast(test.t.b, json BINARY)->Column#15 |
|   └─TableReader_13           | 10000.00 | root      |               | data:TableFullScan_12                                                                                         |
|     └─TableFullScan_12       | 10000.00 | cop[tikv] | table:t1      | keep order:false, stats:pseudo                                                                                |
+------------------------------+----------+-----------+---------------+---------------------------------------------------------------------------------------------------------------+
7 rows in set, 1 warning (0.00 sec)

```

**Review by rebelice — APPROVED** [review]

**Comment by XuHuaiyu** [comment]
The root cause of this problem is that we should NOT enable the ParseAsJsonFlag on a JSON column. (refer: https://github.com/pingcap/tidb/pull/8564)
Can we reset the flag of the casted column when extracting the function into an extra projection?

**Review by winoros — APPROVED** [review]

**Comment by Yisaer** [comment]
> The root cause of this problem is that we should NOT enable the ParseAsJsonFlag on a JSON column. (refer: #8564) Can we reset the flag of the casted column when extracting the function into an extra projection?

updated.

File: planner/core/rule_predicate_push_down.go [hunk 1]
```diff
@@ -292,6 +292,10 @@ func (p *LogicalProjection) appendExpr(expr expression.Expression) *expression.C
 	}
```
Comment (XuHuaiyu) [inline planner/core/rule_predicate_push_down.go:292]: expr.GetType().Clone() ?

Comment (Yisaer) [reply planner/core/rule_predicate_push_down.go:292]: updated.

**Pull request merged**